### PR TITLE
Add more dartdoc to rendering.dart

### PR DESCRIPTION
**Pull request by abarth**
There are still many more left before rendering.dart is complete.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

**Comment by abarth** [comment]
947 `->` 801

File: packages/flutter/lib/src/rendering/auto_layout.dart [hunk 1]
```diff
@@ -11,29 +11,38 @@ import 'object.dart';
 /// for constraints. Also sets up and manages implicit constraints and edit
 /// variables.
 class AutoLayoutRect {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/auto_layout.dart]: not necessarily in this PR, but the various members in this class could do with sample code showing how to use them.

File: packages/flutter/lib/src/rendering/auto_layout.dart [hunk 2]
```diff
@@ -106,40 +118,65 @@ class AutoLayoutParentData extends ContainerBoxParentDataMixin<RenderBox> {
   /// may return null.
   List<al.Constraint> _constructImplicitConstraints() {
     return <al.Constraint>[
-      _rect._left >= al.cm(0.0), // The left edge must be positive.
-      _rect._right >= _rect._left, // Width must be positive.
+      _rect.left >= al.cm(0.0), // The left edge must be positive.
+      _rect.right >= _rect.left, // Width must be positive.
       // Why don't we need something similar for the top and the bottom?
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/auto_layout.dart]: can you replace this comment with a positive TODO, like "TODO(chinmay): check whether we need something similar for the top and bottom"?

File: packages/flutter/lib/src/rendering/auto_layout.dart [hunk 3]
```diff
@@ -106,40 +118,65 @@ class AutoLayoutParentData extends ContainerBoxParentDataMixin<RenderBox> {
...
     ];
   }
 }
 
+/// An abstract interface to control the layout of a [RenderAutoLayout].
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/auto_layout.dart]: should have some comment about it being something you subclass, as opposed to a framework implementation detail

File: packages/flutter/lib/src/rendering/box.dart [hunk 4]
```diff
@@ -996,6 +1009,8 @@ abstract class RenderBox extends RenderObject {
       return true;
     });
   }
+
+  /// In debug mode, paints a border around this render box.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: mention that this is called for every render box if debugPaintSizeEnabled is true

File: packages/flutter/lib/src/rendering/box.dart [hunk 5]
```diff
@@ -1006,6 +1021,8 @@ abstract class RenderBox extends RenderObject {
       return true;
     });
   }
+
+  /// In debug mode, paints a line for each baseline.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: mention debugPaintBaselinesEnabled

File: packages/flutter/lib/src/rendering/box.dart [hunk 6]
```diff
@@ -1033,6 +1050,8 @@ abstract class RenderBox extends RenderObject {
       return true;
     });
   }
+
+  /// In debug mode, paints a rectangle if this render box has received more pointer downs than pointer up events.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: mention the debug flag for this one too

File: packages/flutter/lib/src/rendering/box.dart [hunk 7]
```diff
@@ -1127,8 +1146,9 @@ abstract class RenderBoxContainerDefaultsMixin<ChildType extends RenderBox, Pare
     }
   }
 
+  /// Returns a list containing the children of this render object.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: mention that this isn't especially cheap (why do we even have this?)

File: packages/flutter/lib/src/rendering/editable_line.dart [hunk 8]
```diff
@@ -56,7 +62,10 @@ class RenderEditableLine extends RenderBox {
       ..onLongPress = _handleLongPress;
   }
 
+  /// Called when the selectionc changes.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable_line.dart]: typo

File: packages/flutter/lib/src/rendering/editable_line.dart [hunk 9]
```diff
@@ -56,7 +62,10 @@ class RenderEditableLine extends RenderBox {
...
   SelectionChangedHandler onSelectionChanged;
+
+  /// Called when the inner or outer dimensions of this render objet change.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable_line.dart]: typo

File: packages/flutter/lib/src/rendering/editable_line.dart [hunk 10]
```diff
@@ -101,6 +113,7 @@ class RenderEditableLine extends RenderBox {
 
   List<ui.TextBox> _selectionRects;
 
+  /// The region of text that is selected.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable_line.dart]: if any

File: packages/flutter/lib/src/rendering/editable_line.dart [hunk 11]
```diff
@@ -120,6 +138,13 @@ class RenderEditableLine extends RenderBox {
     markNeedsPaint();
   }
 
+  /// Returns the global coordinates of the endpoints of the given selection.
+  ///
+  /// If the selection is collpsed (and therefore occupies a single point), the
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable_line.dart]: typo

File: packages/flutter/lib/src/rendering/editable_line.dart [hunk 12]
```diff
@@ -120,6 +138,13 @@ class RenderEditableLine extends RenderBox {
...
+  /// returned list is of length one. Otherwise, the selection is not collapsed
+  /// and the returned list is of length two. In this case, however, the two
+  /// points might actually be co-located (e.g., because of a bidirectional
+  /// selection that contains some text but whose ends meet in the middle).
   List<TextSelectionPoint> getEndpointsForSelection(TextSelection selection) {
     _textPainter.layout(); // TODO(mpcomplete): is this hacky?
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable_line.dart]: answer the TODO

File: packages/flutter/lib/src/rendering/flex.dart [hunk 13]
```diff
@@ -90,7 +90,10 @@ typedef double _ChildSizingFunction(RenderBox child, BoxConstraints constraints)
 /// positioned along the cross axis according to the [crossAxisAlignment] property.
 class RenderFlex extends RenderBox with ContainerRenderObjectMixin<RenderBox, FlexParentData>,
                                         RenderBoxContainerDefaultsMixin<RenderBox, FlexParentData> {
-
+  /// Creates a flex render object.
+  ///
+  /// By default, the flex layout is horizontal, aligned to the start of the
+  /// main axis and the center of the cross axis.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/flex.dart]: is the layout aligned that way, or are the children aligned that way?

File: packages/flutter/lib/src/rendering/layer.dart [hunk 14]
```diff
@@ -73,8 +73,12 @@ abstract class Layer {
   @override
   String toString() => '$runtimeType';
 
+  /// The object responsible for creating this layer.
+  ///
+  /// Inherited from [RenderObject.debugCreator]. Used in debug messages.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/layer.dart]: "Inherited" is a bit of a loaded term to use here

File: packages/flutter/lib/src/rendering/layer.dart [hunk 15]
```diff
@@ -109,8 +115,15 @@ class PictureLayer extends Layer {
   }
 }
 
+/// (mojo-only) A layer that represents content from another process.
 class ChildSceneLayer extends Layer {
-  ChildSceneLayer({ this.offset, this.devicePixelRatio, this.physicalWidth, this.physicalHeight, this.sceneToken });
+  ChildSceneLayer({
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/layer.dart]: no constructor dartdoc? :-)

File: packages/flutter/lib/src/rendering/layer.dart [hunk 16]
```diff
@@ -151,9 +165,13 @@ class PerformanceOverlayLayer extends Layer {
...
   Rect overlayRect;
 
-  /// A mask specifying the statistics to display.
+  /// The mask is created by shifting 1 by the index of the specific
+  /// [PerformanceOverlayOption] to enable.
   final int optionsMask;
 
+  /// The rasterizer threshold is an integer specifying the number of frame
+  /// intervals that the rasterizer must miss before it decides that the frame
+  /// is suitable for capturing an SkPicture trace for further analysis.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/layer.dart]: we should document where the SkPicture trace goes and how to use it.

File: packages/flutter/lib/src/rendering/list.dart [hunk 17]
```diff
@@ -11,7 +11,20 @@ import 'viewport.dart';
...
 
+/// A linear layout of children intended for use as a virtual viewport.
+///
+/// Children are layout out in order along the main axis. If [itemExtent] is
+/// non-null, each child is required to have exactly [itemExtent] extent in the
+/// main axis. If [itemExtent] is null, each child is required to have the same
+/// extent in the main axis as the list itself.
+///
+/// In the cross axis, the render list expands to fill the available space and
+/// each child is required to have the same extent in the main axis as the list
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/list.dart]: i think you mean cross axis here

File: packages/flutter/lib/src/rendering/list.dart [hunk 18]
```diff
@@ -35,16 +48,20 @@ class RenderList extends RenderVirtualViewport<ListParentData> {
     addAll(children);
   }
 
+  /// The cross-axis extent of each item in the list.
+  ///
+  /// If [itemExtent] is null, the items are required to match the cross-axis
+  /// extent of the list itself.
   double get itemExtent => _itemExtent;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/list.dart]: pretty sure itemExtent is main axis, not cross axis.

File: packages/flutter/lib/src/rendering/performance_overlay.dart [hunk 19]
```diff
@@ -57,6 +57,9 @@ class RenderPerformanceOverlay extends RenderBox {
     markNeedsPaint();
   }
 
+  /// The rasterizer threshold is an integer specifying the number of frame
+  /// intervals that the rasterizer must miss before it decides that the frame
+  /// is suitable for capturing an SkPicture trace for further analysis.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/performance_overlay.dart]: point to somewhere useful for docs about what this means

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 20]
```diff
@@ -126,24 +161,31 @@ class SemanticsNode extends AbstractNode {
...
 
+  /// Whether this node responds to horizontal scrolling.
   bool get canBeScrolledHorizontally => _flags[_SemanticFlags.canBeScrolledHorizontally];
   set canBeScrolledHorizontally(bool value) => _setFlag(_SemanticFlags.canBeScrolledHorizontally, value, needsHandler: true);
 
+  /// Whether this node responds to vertical scrolling.
   bool get canBeScrolledVertically => _flags[_SemanticFlags.canBeScrolledVertically];
   set canBeScrolledVertically(bool value) => _setFlag(_SemanticFlags.canBeScrolledVertically, value, needsHandler: true);
 
+  /// Whether this node has Boolean state that can be toggled on and off.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: specifically, whether it has a boolean state that can in principle be changed by user interaction.

As opposed to the boolean state "is enabled", for instance.

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 21]
```diff
@@ -154,6 +196,7 @@ class SemanticsNode extends AbstractNode {
     }
   }
 
+  /// Restore this node to its original state.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: default state may be better

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 22]
```diff
@@ -364,6 +432,7 @@ class SemanticsNode extends AbstractNode {
     _listeners.add(listener);
   }
 
+  /// Uploads the semantics tree to the listeners registred with [addListener].
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: typo

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 23]
```diff
@@ -476,39 +546,41 @@ class SemanticsNode extends AbstractNode {
   }
 }
 
+/// Exposes the [SemanticsNode] tree to the underlying platform.
 class SemanticsServer extends mojom.SemanticsServer {
   @override
   void addSemanticsListener(mojom.SemanticsListenerProxy listener) {
+    // TODO(abarth): We should remove the listener when this pipe closes.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: yeah. there's a bug on this.

Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: oh actually the bug is about something related, namely that we should shut down semantics processing while we have no listeners. but similar idea.

File: packages/flutter/lib/src/rendering/stack.dart [hunk 24]
```diff
@@ -49,6 +49,7 @@ class RelativeRect {
     );
   }
 
+  /// A relative rect that covers the entire rect.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/stack.dart]: covers which entire rect?

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/rendering/box.dart (hunk 7 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: It's used by Flow, which needs random access to the children.

File: packages/flutter/lib/src/rendering/layer.dart (hunk 14 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/layer.dart]: Changed to talk about defaults.

File: packages/flutter/lib/src/rendering/layer.dart (hunk 15 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/layer.dart]: I didn't doc the members either.  I'll have to come back to it.

File: packages/flutter/lib/src/rendering/layer.dart (hunk 16 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/layer.dart]: I'm not sure this feature makes sense anymore.  We should replace it with something that integrates with the diagnostic server.

File: packages/flutter/lib/src/rendering/list.dart (hunk 17 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/list.dart]: Yeah.

File: packages/flutter/lib/src/rendering/list.dart (hunk 18 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/list.dart]: Yeah, I got flipped around here.

File: packages/flutter/lib/src/rendering/semantics.dart (hunk 20 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/semantics.dart]: Yeah, that's where I was going with `toggled`, but your formulation is clearer.

File: packages/flutter/lib/src/rendering/semantics.dart (hunk 23 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/semantics.dart]: I added a link to the bug.

File: packages/flutter/lib/src/rendering/stack.dart (hunk 24 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/stack.dart]: Changed to `container`, which is defined terminology for this class (according to the class-level docs).

**Pull request merged**